server/src-lib/Hasura/RQL/DDL/Permission.hs

```diff
@@ -236,16 +235,7 @@ buildSelPermInfo role tabInfo sp = do
 
   -- validate computed fields
   withPathK "computed_fields" $ indexedForM_ computedFields $ \name -> do
-    computedFieldInfo <- askComputedFieldInfo fieldInfoMap name
-    case _cfiReturnType computedFieldInfo of
-      CFRScalar _ -> pure ()
-      CFRSetofTable returnTable -> do
-        returnTableInfo <- askTabInfo returnTable
-        let function = _cffName $ _cfiFunction $ computedFieldInfo
-            errModifier e = "computed field " <> name <<> " executes function "
-                             <> function <<> " which returns set of table "
-                             <> returnTable <<> "; " <> e
-        void $ modifyErr errModifier $ askPermInfo returnTableInfo role PASelect
+    void $ askComputedFieldInfo fieldInfoMap name
 
```

Comment:
Why these checks are removed?

Defining select permission of a table depends upon the existence of select permission on another table. It complicates the adding permissions to metadata in `replace_metadata` query API. The order of permissions in query payload matters now. The `replace_metadata` should pass whatever the order the metadata items are given. Hence the check for select permission are being removed.

While generating the schema for a role, the select permission for computed field returning table is taken into account and omitted if select permission is not defined. However, this logic is been there.

Reply:
> The order of permissions in query payload matters now.

Does this have any potential to make `export_metadata` + `replace_metadata` non-idempotent? That is, do we take care to export the metadata in the right order so that it will never cause any problems when passed directly to `replace_metadata`?

Reply:
> Does this have any potential to make export_metadata + replace_metadata non-idempotent?

Yes. 

> do we take care to export the metadata in the right order so that it will never cause any problems when passed directly to replace_metadata?

No, neither the export metadata nor replace metadata should care about the order. However, in https://github.com/hasura/graphql-engine/pull/3230 we’re making the order deterministic to help CLI to compare and show “diff” of exported metadata.

Reply:
> No, neither the export metadata nor replace metadata should care about the order.

I’m confused. You said

> The order of permissions in query payload matters now. The `replace_metadata` should pass whatever the order the metadata items are given.

which sounds like `replace_metadata` caring about the order to me… but maybe I misunderstood your comment. I interpreted it to mean “we used to do some complicated check to ensure the consistency between a recursive group of things, but now we do a simpler check by assuming we’re given things in the right order so that we can check them top to bottom.” Was my interpretation wrong?

Reply:
Apologies for the unclear comment. 

> The order of permissions in query payload matters now. 

I mean the above sentence if checking for select permissions on returning table. I updated the comment. Hope this is clear to you.

Reply:
Trying to remove all confusions.

We used to check for select permissions on returning tables of computed fields. This complicates the `replace_metadata` logic since the addition of select permission depends on the existence of another select permission. Thus the `replace_metadata` expects the permissions to be in "order". 

Now, the validation of computed fields in permission is simplified by avoiding the select permission check on returning tables.

Reply:
In commit https://github.com/hasura/graphql-engine/pull/3211/commits/c776fe6a2fca490dd8b978e1438e8142a8121639, I've made change to only accept the computed fields returning scalar types in select permissions.